### PR TITLE
factorio: better mod support

### PR DESCRIPTION
**Pull request by YellowOnion**
###### Description of changes

1. Allow us to provide updated versions via a customJson file.
2. Factorio fires an error message about unwritable mod settings as it's in the nix-store, this tweak lets us provide a dat file to allow custom settings.

I also have an ugly script for building a mod listed here: https://github.com/YellowOnion/factorio-mods that I tested it against: https://github.com/YellowOnion/nixos-config/blob/factorio-testing/selene.nix

@Baughn @elitak @erictapen @priegger @lukegb @leifhelm

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [] (Package updates) Added a release notes entry if the change is major or breaking
  - [] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Baughn — APPROVED** [review]
Looks fine to me. I am a bit curious how this would be used in practice, however; I change my mod setup far too often to bother freezing it.

**Comment by YellowOnion** [comment]
@Baughn the selene.nix file is running on my 22.05 VPS (if you want to see it in action), I'm planning on setting up a GitHub Action to auto update the mod.nix file in my repo, the goal here is to make it easier to keep the package up to date, and make a full modded server setup easier. It's impossible to override the current package to keep it up to date between nixpkg merges (factorio is 10 versions out of date on 22.05), or I need some obscure nix feature I'm unaware of to override the variable in the let clause.

I think "leaf" packages like this should be aggressively rolling release even on stable nixos, since breaking changes aren't going to effect downstream packages, and the stale package is more likely to cause compatibility issues with clients who use Steam for example, But this is a good compromise option in the mean time.

**Comment by YellowOnion** [comment]
Can I get more reviews please? it's been almost a month, and the changes aren't all that drastic. Are the others even maintaining the package anymore?

File: pkgs/games/factorio/utils.nix [hunk 1]
```diff
@@ -18,7 +18,9 @@ with lib;
           # NB: there will only ever be a single zip file in each mod derivation's output dir
           ln -s $modDrv/*.zip $out
         done
-      '';
+      '' + (if modsDatFile != null then ''
```
Comment (lukegb) [inline pkgs/games/factorio/utils.nix:21]: Possibly lib.optionalString would make this slightly nicer but on the whole I don't mind either way.

File: nixos/modules/services/games/factorio.nix [hunk 2]
```diff
@@ -136,6 +136,13 @@ in
           derivations via nixos-channel. Until then, this is for experts only.
         '';
       };
+      mods-dat = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = lib.mdDoc ''
+        Mods settings can be changed by specifying a dat file.
```
Comment (lukegb) [inline nixos/modules/services/games/factorio.nix:143]: Indent slightly more, and if possible include a link to what this is. It's not obvious from reading this description.

**Review by lukegb — CHANGES_REQUESTED** [review]
Sorry it's taken a while: I find it very difficult in the GitHub UI to find reviews that are still pending my attention.

A few nits but nothing major - take a look at them and I'll merge.

**Pull request merged**